File: python/00523_continuous_subarray_sum.py
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        mod_map = {0: -1}  # remainder -> earliest index
        prefix_sum = 0

        for i, num in enumerate(nums):
            prefix_sum += num
            mod = prefix_sum % k if k != 0 else prefix_sum

            if mod in mod_map:
                if i - mod_map[mod] > 1:  # length at least 2
                    return True
            else:
                mod_map[mod] = i

        return False


# Time complexity - O(n)
# Space complexity - O(min(n, k))

File: java/00523_continuous_subarray_sum.java
public class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        Map<Integer, Integer> modMap = new HashMap<>();
        modMap.put(0, -1);

        int prefixSum = 0;

        for (int i = 0; i < nums.length; i++) {
            prefixSum += nums[i];

            int mod = k != 0 ? prefixSum % k : prefixSum;

            if (modMap.containsKey(mod)) {
                if (i - modMap.get(mod) > 1) {
                    return true;
                }
            } else {
                modMap.put(mod, i);
            }
        }

        return false;
    }
}

// Time complexity - O(n)
// Space complexity - O(min(n, k))